### PR TITLE
Importable packages and reload

### PR DESCRIPTION
**Pull request by sjrct**
This PR changes the way that imports are done by creating a package called `halibot.packages` from which they can be directly imported. This also allows the use of `importlib.reload`.

Reload is implemented by reloading the python modules that are being used by running halibot modules and are within the package specified to reload. This implies that packages imported by halibot modules are not reloaded and must be reload explicitly if need be.

There is one potential issue that this introduces, which is that the import path for `halibot.packages` is singular while the `package-path` is dependent on the `Halibot` class instance, which is potentially not singular. Currently, however, multiple `Halibot` instances are not created within the same python interpreter, so this is not as of yet a problem, and I believe the elegance of being able to use the import statement is benefit enough to forego that.

### CODE REVIEW

File: halibot/halibot.py [hunk 1]
```diff
@@ -120,8 +120,18 @@ def get_package(self, name):
...
 	# TODO: Reload a class, and restart all modules of that class
-	def reload(self, cls):
-		pass
+	def reload(self, name):
+		mod = importlib.reload(self.get_package(name))
+
+		parent = 'halibot.packages.' + name
+		for k,o in self.objects.items():
+			if o.__module__.startswith(parent + '.'):
+				o.shutdown()
```
Comment (richteer) [inline halibot/halibot.py]: This should probably be `o._shutdown()`, so the eventloop and thread are cleanly removed.

Maybe we should rename one of the `._?shutdown()` method to something more descriptive...

File: halibot/halibot.py [hunk 2]
```diff
@@ -5,7 +5,8 @@
 import json
 import threading
 import os, sys
-from importlib.machinery import SourceFileLoader
+import importlib
+import halibot.packages
```
Comment (richteer) [inline halibot/halibot.py:9]: Do we need to modify `sys.path` for this at all, or are we to assume that `halibot` exists already in the python lib path?

File: halibot/halibot.py [hunk 3]
```diff
@@ -72,6 +73,7 @@ def add_instance(self, name, inst):
 	def _load_config(self):
 		with open("config.json","r") as f:
 			self.config = json.loads(f.read())
+			halibot.packages.__path__ = self.config.get("package-path", [])
```
Comment (richteer) [inline halibot/halibot.py:76]: To make sure I'm understanding this, this won't change what needs to go in `package-path`, right? Those semantics should be the same?

File: halibot/halibot.py [hunk 4]
```diff
@@ -115,11 +117,7 @@ def _instantiate_objects(self, key):
 			self.add_instance(k, obj(self, conf))
 
 	def get_package(self, name):
-		for prefix in self.config['package-path']:
-			path = os.path.join(prefix, name, '__init__.py')
-			if os.path.exists(path):
-				return SourceFileLoader(path, path).load_module()
-		return None
+		return importlib.import_module('halibot.packages.' + name)
```
Comment (richteer) [inline halibot/halibot.py:120]: So much cleaner 👍

**Review by richteer — COMMENTED** [review]
LGTM (mostly probably)

File: halibot/halibot.py (hunk 2 above)
Comment (sjrct) [reply halibot/halibot.py:9]: All the modules/agents assume it does, so I think it is reasonable to assume so here as well.

File: halibot/halibot.py (hunk 1 above)
Comment (sjrct) [reply halibot/halibot.py]: You're right, should be `_shutdown`

File: halibot/halibot.py (hunk 3 above)
Comment (sjrct) [reply halibot/halibot.py:76]: Correct, the `package-path` remains unchanged from previous versions.

Comment (richteer) [reply halibot/halibot.py:76]: Good good, was afraid we'd have to change defaults again

File: halibot/halibot.py (hunk 2 above)
Comment (richteer) [reply halibot/halibot.py:9]: Just tested - works as expecting. Not sure if our tests use the local directory, but I still do in my own tests. Still works that way, which is nice for lazy me.

**Pull request merged**